### DISCUSSION TITLE
Class-based config deprecation warning (fields usage)

### DISCUSSION THREAD

**Discussion by deivydaslipskis** [pydantic/pydantic#7076] [Question] [answered]
Hi, I am migrating from Pydantic v1 to v2 and receiving warnings like these:
1st:
```
PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.1.1/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)
```
    
2nd:
```
UserWarning: Valid config keys have changed in V2:
  * 'fields' has been removed
    warnings.warn(message, UserWarning)
```
   
Here is how I used class-based config so far:

1st simple example:

```
from pydantic_settings import BaseSettings

class ProjSettings(BaseSettings):
    var1: str
    var2: str

    class Config:
        env_file = ".env"
```
    
2nd example using fields:

```
from pydantic_settings import BaseSettings

class DBSettings(BaseSettings):
    var1: str = "test"
    var2: bool = True

    run_path: str | None = None
    out_path: str | None = None
    file_path: str | None = None

    class Config:
        fields = {
            "run_path": {"exclude": True},
            "out_path": {"exclude": True},
            "file_path": {"exclude": True},
        }    
```

How would I need to change these lines of code to not receive these warnings? I know that these are just warnings and I could keep it as is, but I want to make it correct.

I read about config changes - https://docs.pydantic.dev/dev-v2/migration/#changes-to-config but I am still not sure how to make the changes.

**Comment by deivydaslipskis** [comment]
I think I should use `model_config` https://docs.pydantic.dev/2.1/usage/model_config but I am not sure to pass `env_file` to it and I could not find an example in the documentation (especially don't know how to remake `fields`). Perhaps someone has an example or explanation? Thank you!

**Comment by deivydaslipskis** [comment]
I think I have figured out how to use .env files. This seems to work:

```
from pydantic_settings import BaseSettings, SettingsConfigDict

class ProjSettings(BaseSettings):
    model_config = SettingsConfigDict(env_file=".env")
    var1: str
    var2: str

```

But I still don't know what to do with `fields`. An example of how to use `fields` or something similar in Pydantic v2 would really help, thanks in advance!

**Reply by Viicos** [reply, marked as answer]
```python
from typing import Annotated, TypeVar

from pydantic import Field
from pydantic_settings import BaseSettings

T = TypeVar("T")

ExcludedField = Annotated[T, Field(exclude=True)]


class DBSettings(BaseSettings):
    var1: str = "test"
    var2: bool = True

    run_path: ExcludedField[str | None] = None
    out_path: ExcludedField[str | None] = None
    file_path: ExcludedField[str | None] = None
```

**Reply by deivydaslipskis** [reply]
Thank you! May I ask what is the purpose of `T = TypeVar("T")`?

**Reply by Viicos** [reply]
You can find more here: [`TypeVar`](https://docs.python.org/3/library/typing.html#typing.TypeVar). It is normally used for generics.

In this context, it basically acts as an intermediary type/template, in fact you could do the following (equivalent):

```python
def make_annotated(field_type):
    return Annotated[field_type, Field(...)]


class Model(BaseSettings):
    my_field: make_annotated(str | None)
```

But type checkers aren't usually happy with having function calls as type hints, that's why it's better to do it as in https://github.com/pydantic/pydantic/discussions/7076#discussioncomment-6699138.

In our case, `ExcludedField` is "incomplete" as is, it still needs to be parametrized with a type, and you can do that using `[]` (e.g. `ExcludedField[str | None]`).

By the way, [since Python 3.12](https://docs.python.org/3.12/reference/compound_stmts.html#generic-type-aliases), you will be able to do the following:

```python
type ExcludedField[T] = Annotated[T, Field(exclude=True)]   # No need to use any TypeVars
```

The way generics work have been improved (see https://docs.python.org/3.12/library/typing.html#generics), and gets closer to what you can see in other languages (e.g. [TypeScript](https://www.typescriptlang.org/docs/handbook/2/generics.html))

**Reply by agn-7** [reply]
despite using `model_config = SettingsConfigDict(env_file=".env", case_sensitive=True)` rather than the following code snippet I still get the same warning:

```
    class Config:
        env_file = ".env"
        case_sensitive = True
```

Warning:

```
Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.4/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)
```

**Reply by Viicos** [reply]
You are most probably still using the old `Config` somewhere.

**Reply by agn-7** [reply]
Oh, good catch, thank you. I just saw somewhere I also had `BaseModel` with config-class like this:

```
class MessageCreate(BaseModel):
    role: str = "human"
    content: str

    class Config:
        from_attributes = True
```

And I fixed that as follows:

```
class MessageCreate(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    role: str = "human"
    content: str
```

Is it right? Because I couldn't find `from_attributes=True` params in docs migration part.

**Reply by Viicos** [reply]
Yes, see here: https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.from_attributes
